File: bin/chess/game/pieces/king.py
from typing import Any
from utils.functions import get_window_pos, get_game_pos
import pygame
from .piece import Piece
from board_utils.square import Square
from chess.CONSTANTS import WHITE, BLACK
from utils.types import WindowPosition, GamePosition, Positions, Squares


class King(Piece):

    def __init__(self, file: str, rank: int, color: pygame.Color, min_x: int, max_x: int, min_y: int, max_y: int, square_width: int, square_height: int, win_width: int, win_height: int) -> None:
        pygame.init()       # initialize pygame
        super().__init__(file, rank, 'King', color)
        self.min_x = min_x
        self.max_x = max_x
        self.min_y = min_y
        self.max_y = max_y
        self.square_width = square_width       # width of a single square position
        self.square_height = square_height        # height of single square position
        self.win_width = win_width       # width of the window
        self.win_height = win_height      # height of the window
        self.x, self.y = self.piece_x, self.piece_y = get_window_pos(
            self.file, self.rank, self.possible_files, self.square_width, self.square_height)
        self.attacked_pieces = []      # list of pieces being attacked by self
        self.attackers = []      # list of pieces attacking self
        self.has_moved = False

    def validate(self, position: GamePosition, squares: Squares) -> bool:
        file, rank = position
        possible_distances = [0, 1]
        if abs(rank-self.rank) in possible_distances and abs(ord(file)-ord(self.file)) in possible_distances:
            piece_color = squares[(file+str(rank))].piece.color
            if piece_color == self.color:
                return False
            else:
                squares[file+str(rank)].piece, squares[self.file+str(self.rank)
                                                       ].piece = squares[self.file+str(self.rank)].piece, squares[file+str(rank)].piece
                return self.check(position=position, squares=squares)
        return False

    def check(self, pieces: list[Piece] = [], position: GamePosition = tuple(), squares: dict[str, Square] = dict()) -> bool:
        if position and squares:
            if squares[position[0]+str(position[1])].attacked:
                return True
            return False
        else:
            for piece_list in pieces:
                for piece in piece_list:
                    if self.attackers or (self.x, self.y) in piece.attacked_pieces:
                        return True
            return False

    def get_possible_positions_from_current_position(self, position: GamePosition, squares: Squares) -> Positions:
        file, rank = position
        prev_rank = rank-1 if rank > 1 else None
        next_rank = rank+1 if rank < 8 else None
        prev_file = self.possible_files[self.possible_files.index(
            file)-1] if file != self.possible_files[0] else None
        next_file = self.possible_files[self.possible_files.index(
            file)+1] if file != self.possible_files[-1] else None
        return list(filter((lambda i: ((squares[i[0]+str(i[1])].piece.color != self.color) and (all(i)))), [(prev_file, prev_rank), (prev_file, rank), (prev_file, next_rank), (file, prev_rank), (file, next_rank), (next_file, prev_rank), (next_file, rank), (next_file, next_rank)]))

    def checkmate(self, pieces: list[Piece], squares: Squares) -> bool:
        if self.check(pieces):
            possible_positions = self.get_possible_positions_from_current_position(
                (self.file, self.rank), squares).append((self.file, self.rank))
            filtered_possible_positions = list(
                filter(lambda i: self.check(pieces, i, self.squares), possible_positions))
            if len(possible_positions) == len(filtered_possible_positions):
                return True
            return False
        return False

    def castle(self, rook, long_castle: bool, squares):
        if long_castle:
            if squares['D'+str(self.rank)].empty and squares['C'+str(self.rank)].empty and squares['B'+str(self.rank)].empty:
                if not self.has_moved and not rook.has_moved:
                    if not (self.check(position=(self.file, self.rank), squares=squares) or self.check(position=('D', self.rank), squares=squares) or self.check(position=('C', self.rank), squares=squares) or self.check(position=('B', self.rank), squares=squares)):
                        self.file = 'C'
                        rook.file = 'D'
                        return True
        else:
            if squares['F'+str(self.file)].empty or squares['G'+str(self.file)].empty:
                if not self.has_moved and not rook.has_moved:
                    if not (self.check(position=(self.file, self.rank), squares=squares) or self.check(position=('F', self.rank), squares=squares) or self.check(position=('G', self.rank), squares=squares)):
                        self.file = 'G'
                        rook.file = 'F'
                        return True
        return False

File: src/chess/game/pieces/King.java
package game.pieces;

// Import Java modules
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Collection;
import java.awt.Color;
import game.boardUtils.Square;
import game.utils.Constants;
import game.utils.FileC;
import game.utils.RankC;
import game.utils.UtilMethods;

public class King extends Piece {

    private ArrayList<Square> attackers;
    private boolean hasMoved;

    public King(FileC file, RankC rank, Color color) {
        super("King", file,rank,color);
        attackedPieces = new ArrayList<>();
        attackers = new ArrayList<>();
        hasMoved = false;
    }

    public ArrayList<Square> getAttackers() {
        return(attackers);
    }

    public boolean hasMoved() {
        return(hasMoved);
    }

    @Override
    public ArrayList<Square> getPossiblePositions(String pos, Collection<Square> squares) {
        FileC file = FileC.valueOf(pos.substring(0,1));
        RankC rank = RankC.valueOf(pos.substring(1));
        ArrayList<Square> possiblePositions = new ArrayList<>(63);
        Square topLeftSquare = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) file.file())-1))), RankC.valueOf(Integer.toString(rank.rank()-1)));
        if (topLeftSquare != null) possiblePositions.add(topLeftSquare);
        Square leftSquare = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) file.file())-1))), rank);
        if (leftSquare != null) possiblePositions.add(leftSquare);
        Square bottomLeftSquare = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) file.file())-1))), RankC.valueOf(Integer.toString(rank.rank()+1)));
        if (bottomLeftSquare != null) possiblePositions.add(bottomLeftSquare);
        Square topSquare = this.getSquare(squares, file, RankC.valueOf(Integer.toString(rank.rank()-1)));
        if (topSquare != null) possiblePositions.add(topSquare);
        Square bottomSquare = this.getSquare(squares, file, RankC.valueOf(Integer.toString(rank.rank()+1)));
        if (bottomSquare != null) possiblePositions.add(bottomSquare);
        Square topRightSquare = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) file.file())+1))), RankC.valueOf(Integer.toString(rank.rank()-1)));
        if (topRightSquare != null) possiblePositions.add(topRightSquare);
        Square rightSquare = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) file.file())+1))), rank);
        if (rightSquare != null) possiblePositions.add(rightSquare);
        Square bottomRightSquare = this.getSquare(squares, FileC.valueOf(Character.toString((char) (((int) file.file())+1))), RankC.valueOf(Integer.toString(rank.rank()+1)));
        if (bottomRightSquare != null) possiblePositions.add(bottomRightSquare);
        return(possiblePositions);
    }

    @Override
    public boolean validate(FileC f, RankC r, Collection<Square> squares, King king) {
        int[] pos = UtilMethods.getWindowPos(this.getFile(), this.getRank(), Constants.SQUARE_WIDTH, Constants.SQUARE_HEIGHT);
        ArrayList<Square> possiblePositions = getPossiblePositions(UtilMethods.getGamePos(pos[0], pos[1], Constants.SQUARE_WIDTH, Constants.SQUARE_HEIGHT), squares);
        Square s = this.getSquare(squares,f,r);
        if (s != null && !s.getAttacked() && s.getPieceColor() != this.getColor()) {
            return(possiblePositions.contains(s) && !king.check(f, r, squares));
        }
        return(false);
    }

    public boolean validate(FileC f, RankC r, Collection<Square> squares) {
        return(this.validate(f, r, squares, this));
    }

    public boolean check(FileC f, RankC r,Collection<Square>  squares) {
        if (f == null || r == null || squares == null) {
            throw new UnsupportedOperationException("The 'f', 'r' or 'squares' parameters cannot be 'null'");
        } else {
            Square s = this.getSquare(squares, f, r);
            if (s != null && s.getAttacked()) {
                return(true);
            }
            return(false);
        }
    }

    public boolean check(FileC f, RankC r, ArrayList<ArrayList<Piece>> pieces, Collection<Square> squares) {
        if (f == null || r == null || pieces == null || squares == null) {
            throw new UnsupportedOperationException("The 'f', 'r', 'pieces' or 'squares' parameters cannot be 'null'");
        } else {
            for (ArrayList<Piece> pieceList: pieces) {
                for (Piece piece: pieceList) {
                    if (!attackers.isEmpty() || piece.getAttackedPieces().contains(getSquare(squares,f,r))) {
                        return(true);
                    }
                }
            }
            return(false);
        }
    }

    public boolean checkmate(ArrayList<ArrayList<Piece>> pieces, ArrayList<Square> squares) {
        if (check(getFile(), getRank(), pieces, squares)) {
            int[] pos = UtilMethods.getWindowPos(this.getFile(), this.getRank(), Constants.SQUARE_WIDTH, Constants.SQUARE_HEIGHT);
            ArrayList<Square> possiblePositions = this.getPossiblePositions(UtilMethods.getGamePos(pos[0], pos[1], Constants.SQUARE_WIDTH, Constants.SQUARE_HEIGHT), squares);
            ArrayList<Square> filteredPossiblePositions = new ArrayList<>(possiblePositions.size());
            for (int i=0; i<possiblePositions.size(); i++) {
                Square square = possiblePositions.get(i);
                if (check(square.getFile(), square.getRank(), squares)) {
                    filteredPossiblePositions.add(square);
                }
            }
            if (filteredPossiblePositions.size() == possiblePositions.size()) {
                return(true);
            }
            return(false);
        }
        return(false);
    }

    public boolean castle(Rook rook, boolean longCastle, ArrayList<Square> squares) {
        if (longCastle) {
            if (this.getSquare(squares, FileC.D, this.getRank()).empty() && this.getSquare(squares, FileC.C, this.getRank()).empty() && this.getSquare(squares, FileC.B, this.getRank()).empty()) {
                if (!hasMoved && rook.hasMoved()) {
                    if (!check(getFile(), getRank(), squares) && !check(FileC.D, getRank(), squares) && !check(FileC.C, getRank(), squares) && !check(FileC.B, getRank(), squares)) {
                        setFile(FileC.C);
                        rook.setFile(FileC.D);
                        return(true);
                    }
                    return(false);
                }
                return(false);
            }
            return(false);
        } else {
            if (this.getSquare(squares, FileC.F, this.getRank()).empty() && this.getSquare(squares, FileC.G, this.getRank()).empty()) {
                if (!hasMoved && !rook.hasMoved()) {
                    if (!check(getFile(), getRank(), squares) && !check(FileC.F, getRank(), squares) && !check(FileC.G, getRank(), squares)) {
                        setFile(FileC.G);
                        rook.setFile(FileC.F);
                        return(true);
                    }
                    return(false);
                }
                return(false);
            }
            return(false);
        }
    }

}
